File: chapter3/dda.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void dda(int x0, int y0, int x1, int y1, void (*setPixel)(int, int)) {
    int dx = abs(x1 - x0), dy = abs(y1 - y0);
    int offset = dx > dy ? dx : dy;
    float sx = float(dx) / offset, sy = float(dy) / offset;
    
    float x = x0, y = y0;
    for (int i=0; i<offset; ++i) {
        setPixel(round(x), round(y));
        x += sx;
        y += sy;
    }
}

void print(int x, int y) {
    printf("setPixel %d %d\n", x, y);
}

int main() {
    dda(1, 3, 5, 10, print);
    return 0;
}

File: pyproject/week2/dda.py
from ..common.grid import Grid, Node
from PyQt5 import QtWidgets


def dda(p0, p1, func):
    dx = p1[0] - p0[0]
    dy = p1[1] - p0[1]
    offset = int(abs(dx) if abs(dx) > abs(dy) else abs(dy))
    sx = dx / offset
    sy = dy / offset
    x = p0[0]
    y = p0[1]
    for i in range(offset + 1):
        func(round(x), round(y))
        x = x + sx
        y = y + sy


class DDAGrid(Grid):
    def __init__(self, line_func, **params):
        super().__init__(**params)

        self.line_func = line_func
        self.p = None

    def grid_click(self, node: Node):
        if self.p is None:
            self.p = [node.x, node.y]
        else:
            self.line_func(self.p, [node.x, node.y], self.toggle)
            self.p = None


if __name__ == "__main__":

    import sys

    app = QtWidgets.QApplication(sys.argv)
    gui = DDAGrid(n=10, line_func=dda)
    gui.show()
    sys.exit(app.exec_())
